library/private/options_tools.h.in

```diff
@@ -0,0 +1,275 @@
+  bool b2;
+  std::istringstream(str) >> b1;
+  std::istringstream(str) >> std::boolalpha >> b2;
+  return b1 || b2;
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Parse provided string into an int
+ * Rely on std::stoi for parsing
+ * Can throw options::parsing_exception in case of failure to parse
+ */
+template<>
+int parse(const std::string& str)
+{
+  // TODO implement more parsing possibilities ?
+  try
+  {
+    return std::stoi(str);
+  }
+  catch (std::invalid_argument const&)
+  {
+    throw options::parsing_exception("Cannot parse " + str + " into an int");
+  }
+  catch (std::out_of_range const&)
+  {
+    throw options::parsing_exception(
+      "Cannot parse " + str + " into an int as it would go out of range");
+  }
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Parse provided string into a double
+ * Rely on std::stod for parsing
+ * Can throw options::parsing_exception in case of failure to parse
+ */
+template<>
+double parse(const std::string& str)
+{
+  // TODO implement more parsing possibilities ?
+  try
+  {
+    return std::stod(str);
+  }
+  catch (std::invalid_argument const&)
+  {
+    throw options::parsing_exception("Cannot parse " + str + " into a double");
+  }
+  catch (std::out_of_range const&)
+  {
+    throw options::parsing_exception(
+      "Cannot parse " + str + " into a double as it would go out of range");
+  }
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Parse provided string into a ratio_t
+ * Rely on <double> version for parsing
+ * Can throw options::parsing_exception in case of failure to parse
+ */
+template<>
+ratio_t parse(const std::string& str)
+{
+  // TODO implement proper ratio parsing
+  return options_tools::parse<double>(str);
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Return provided string
+ */
+template<>
+std::string parse(const std::string& str)
+{
+  return str;
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Parse provided string into a double vector
+ * Split the string using ',', then rely on double version for parsing each token
+ * Can throw options::parsing_exception in case of failure to parse
+ */
+template<>
+std::vector<double> parse(const std::string& str)
+{
+  // TODO implement more parsing possibilities, eg different types of tokens
+  std::vector<double> vec;
+  std::istringstream split(str);
+  for (std::string each; std::getline(split, each, ',');)
+  {
+    vec.emplace_back(options_tools::parse<double>(each));
+  }
+  return vec;
+}
+
+// TODO Improve string generation
+//----------------------------------------------------------------------------
+/**
+ * Generic templated string generation from provided value
+ * rely on std::to_string
+ */
+template<typename T>
+std::string format(const T& var)
+{
+  return std::to_string(var);
```

Comment:
maybe add `static_assert(std::is_arithmetic_v<T>, "Cannot format this type")`

Reply:
cant we rely on std::to_string to tell us what it can handle or not ?

Reply:
it's just to have an error more clear to the user if there is no specialization available instead of getting a compilation error related to `std::to_string`

Reply:
the libf3d user cannot get a compilation error here afaics.